### PR TITLE
KAFKA-16033: Commit retry logic fixes

### PR DESCRIPTION
**Pull request by lianetm**
This PR modifies the commit manager for improved retry logic & fixing bugs:
- defines high level functions for each of the different types of commit: commitSync, commitAsync, autoCommitSync (used from consumer close), autoCommitAsync (on interval), autoCommitNow (before revocation).
- moves retry logic to these caller functions, keeping a common response error handling that propagates errors that each caller functions retry as it needs. 

Fixes the following issues:
- auto-commit before revocation should retry with latest consumed offsets
- auto-commit before revocation should only reset the timer once, when the rebalance completes
- StaleMemberEpoch error (fatal) is considered retriable only when committing offsets before revocation, where it is retried with backoff if the member has a valid epoch. All other commits will fail fatally on stale epoch. Note that auto commit on the interval (autoCommitAsync) does not have any specific retry logic for the stale epoch, but will effectively retry on the next interval (as it does for any other fatal error) 
- fix duplicated and noisy logs for auto-commit

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 1]
```diff
@@ -1344,8 +1343,8 @@ public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets, Duration
         long commitStart = time.nanoseconds();
         try {
             Timer requestTimer = time.timer(timeout.toMillis());
-            // Commit with a timer to control how long the request should be retried until it
-            // gets a successful response or non-retriable error.
+            // Commit with a retry timeout (the commit request will be retried until it gets a
+            // successful response, non-retriable error, or the timeout expires)
             CompletableFuture<Void> commitFuture = commit(offsets, true, Optional.of(timeout.toMillis()));
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: While you are at it, I think we can remove the "isWakeupable" parameter and just set the  `wakeUpTrigger` in the calling context.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java [hunk 2]
```diff
@@ -149,8 +148,14 @@ private void process(final CommitApplicationEvent event) {
         }
 
         CommitRequestManager manager = requestManagers.commitRequestManager.get();
-        Optional<Long> expirationTimeMs = event.retryTimeoutMs().map(this::getExpirationTimeForTimeout);
-        event.chain(manager.addOffsetCommitRequest(event.offsets(), expirationTimeMs, false));
+        CompletableFuture<Void> commitResult;
+        if (event.retryTimeoutMs().isPresent()) {
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java]: It's a bit weird that we use `retryTimeoutMs` to carry the information that this is a sync-commit vs asynccommit.

How about going all-in here and just having `AsyncCommitApplicationEvent` and a `SyncCommitApplicationEvent`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 3]
```diff
@@ -1147,15 +1202,15 @@ List<NetworkClientDelegate.UnsentRequest> drain(final long currentTimeMs) {
          * futures with a TimeoutException.
          */
         private void failAndRemoveExpiredCommitRequests(final long currentTimeMs) {
-            unsentOffsetCommits.removeIf(req -> req.maybeExpire(currentTimeMs));
+            unsentOffsetCommits.forEach(req -> req.maybeExpire(currentTimeMs));
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Foreach javadoc says

>     * The behavior of this method is unspecified if the action performs
>     * side-effects that modify the underlying source of elements, unless an
>     * overriding class has specified a concurrent modification policy.

You are removing from `unsentOffsetCommits` inside `maybeExpire`. Are we sure we won't run into ConcurrentModificationExceptions here?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 4]
```diff
@@ -204,126 +205,315 @@ private static long findMinTime(final Collection<? extends RequestState> request
...
+                asyncCommitResult.completeExceptionally(asyncCommitException);
+            } else {
+                asyncCommitResult.complete(null);
+            }
+        });
+        return asyncCommitResult;
+    }
+
+    /**
+     * Commit offsets, retrying on expected retriable errors while the retry timeout hasn't expired.
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:367]:
```suggestion
     * Commit offsets, retrying on expected retriable errors while the retry timeout hasn't expired.
     *
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 5]
```diff
@@ -204,126 +205,315 @@ private static long findMinTime(final Collection<? extends RequestState> request
...
+                groupInstanceId,
+                expirationTimeMs,
+                retryBackoffMs,
+                retryBackoffMaxMs,
+                memberInfo);
+    }
+
+    private void commitSyncWithRetries(OffsetCommitRequestState requestAttempt,
+                                       CompletableFuture<Void> result) {
+        requestAttempt.resetFuture();
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: In the first call we just created a fresh future. Maybe move the `resetFuture` to the `commitSyncWithRetries` call below?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 6]
```diff
@@ -204,126 +205,315 @@ private static long findMinTime(final Collection<? extends RequestState> request
...
+        pendingRequests.addOffsetCommitRequest(requestAttempt);
+
+        // Retry the same commit request while it fails with RetriableException and the retry
+        // timeout hasn't expired.
+        requestAttempt.future.whenComplete((res, error) -> {
+            if (error == null) {
+                result.complete(null);
+            } else {
+                if (error instanceof RetriableException) {
+                    // if (error instanceof TimeoutException && requestAttempt.retryTimeoutExpired
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Do we want to commit that commented out code?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 7]
```diff
@@ -204,126 +205,315 @@ private static long findMinTime(final Collection<? extends RequestState> request
...
+                    // (requestAttempt.lastReceivedMs)) {
+                    if (error instanceof TimeoutException && requestAttempt.isExpired) {
+                        log.debug("OffsetCommit timeout expired so it won't be retried anymore");
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Maybe that's even worth and `info` log?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 8]
```diff
@@ -204,126 +205,315 @@ private static long findMinTime(final Collection<? extends RequestState> request
...
             return CompletableFuture.completedFuture(null);
         }
-        return pendingRequests.addOffsetCommitRequest(offsets, expirationTimeMs, retryOnStaleEpoch).future;
+        OffsetCommitRequestState commitRequest = createOffsetCommitRequest(offsets, Optional.empty());
+        pendingRequests.addOffsetCommitRequest(commitRequest);
+
+        CompletableFuture<Void> asyncCommitResult = new CompletableFuture<>();
+        commitRequest.future.whenComplete((committedOffsets, error) -> {
+            if (error != null) {
+                Throwable asyncCommitException;
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: For `commitSync` you have `commitSyncExceptionForError`, so maybe we want to extract this into `commitAsyncExceptionForError`? Symmetry ftw

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 9]
```diff
@@ -204,126 +205,315 @@ private static long findMinTime(final Collection<? extends RequestState> request
...
+        }
+
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        OffsetCommitRequestState requestState =
+            createOffsetCommitRequest(subscriptions.allConsumed(), Optional.of(retryExpirationTimeMs));
+        maybeAutoCommitSyncNowWithRetries(requestState, result);
+        return result;
+    }
+
+    private void maybeAutoCommitSyncNowWithRetries(OffsetCommitRequestState requestAttempt,
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Why `maybe`? pretty sure we are doing it

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 10]
```diff
@@ -204,126 +205,315 @@ private static long findMinTime(final Collection<? extends RequestState> request
...
+                                                   CompletableFuture<Void> result) {
+        requestAttempt.resetFuture();
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Could be moved down, as in `commitSyncWithRetries`

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 11]
```diff
@@ -719,26 +846,27 @@ NetworkClientDelegate.UnsentRequest buildRequestWithResponseHandling(final Abstr
...
             } catch (Throwable t) {
-                log.error("Unexpected error handling response for ", requestDescription(), t);
+                log.error("Unexpected error handling response for {}", requestDescription(), t);
                 future().completeExceptionally(t);
             }
         }
 
         abstract void onResponse(final ClientResponse response);
+
+        abstract boolean retryTimeoutExpired(long currentTimeMs);
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Considered moving `expirationTimeout` to this super-class and inlining this? After all, expiration timeout is functionality that really belongs with the retry logic

**Review by lucasbru — COMMENTED** [review]
Thanks for the PR! That is a good refactoring, and makes the code a lot clearer. I remember the last time I was confused by the many variations of commits that interleave here, this change cleans it up nicely. 

It's a large change, I cannot say I checked every line, but I left a few comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java (hunk 2 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java]: Say no more, I was leaning towards that too (just a bit hesitant to introduce more on this already big PR), but totally agree that it's conceptually clearer, the result is nice, done.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 1 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Makes sense, done.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java (hunk 5 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Agree. Moved it for this and the others (commitAsync and fetch)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java (hunk 6 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Oops no...removed this and another similar one that was left

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java (hunk 11 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Good point, done. I missed that probably influenced by the fact that, in the case of the commits, the expirationTimeout does not apply to all request. But I do agree with you that the expiration belongs with the retry logic so it does fit well at the RetriableRequest level (just Optional).

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java (hunk 8 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Nice symmetry indeed, done.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java (hunk 3 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Definitely, good catch, fixed.

**Comment by lianetm** [comment]
Thanks for the comments @lucasbru , all addressed.

**Review by lucasbru — COMMENTED** [review]
LGTM, thanks!

**Pull request merged**